File: mac/PabloCompanion/Services/PracticeWebSocketClient.swift
import AVFoundation
import Foundation
import os

/// WebSocket client for practice mode audio streaming.
///
/// Sends therapist mic audio (PCM 16kHz mono) to the backend and receives
/// Pablo Bear's response audio (PCM 24kHz mono). Uses the hybrid text/binary
/// protocol defined in practice-mode-api.md.
final class PracticeWebSocketClient: @unchecked Sendable {
    // MARK: - Public state

    enum ConnectionState: Sendable {
        case disconnected
        case connecting
        case authenticating
        case waitingForSession
        case active
        case ending
    }

    enum PabloState: String, Sendable {
        case listening
        case processing
        case speaking
    }

    // MARK: - Callbacks (set by PracticeViewModel on MainActor)

    var onConnectionStateChanged: (@Sendable (ConnectionState) -> Void)?
    var onPabloStateChanged: (@Sendable (PabloState) -> Void)?
    var onAudioReceived: (@Sendable (Data, Bool) -> Void)?
    var onSessionStarted: (@Sendable (String) -> Void)?
    var onSessionEnded: (@Sendable (Int) -> Void)?
    var onError: (@Sendable (String, Bool) -> Void)?

    // MARK: - Private

    private let logger = Logger(subsystem: AppConstants.appBundleID, category: "PracticeWebSocket")
    private var webSocket: URLSessionWebSocketTask?
    private var urlSession: URLSession?
    private let lock = NSLock()
    private var sendSequence: UInt16 = 0
    private var heartbeatTask: Task<Void, Never>?
    private var receiveTask: Task<Void, Never>?

    private var _state: ConnectionState = .disconnected
    var state: ConnectionState {
        lock.lock()
        defer { lock.unlock() }
        return _state
    }

    private func setState(_ newState: ConnectionState) {
        lock.lock()
        _state = newState
        lock.unlock()
        onConnectionStateChanged?(newState)
    }

    // MARK: - Connect

    func connect(url: URL) {
        disconnect()

        setState(.connecting)

        let session = URLSession(configuration: .default)
        urlSession = session

        let task = session.webSocketTask(with: url)
        webSocket = task
        task.resume()

        setState(.authenticating)
        startReceiving()
    }

    // MARK: - Session lifecycle

    func startSession(sessionId: String) {
        let message: [String: Any] = [
            "type": "session_start",
            "session_id": sessionId,
        ]
        sendJSON(message)
        setState(.waitingForSession)
    }

    func endSession() {
        sendJSON(["type": "session_end"])
        setState(.ending)
    }

    func pauseAudio() {
        sendJSON(["type": "audio_pause"])
    }

    func resumeAudio() {
        sendJSON(["type": "audio_resume"])
    }

    // MARK: - Send audio

    /// Sends a 20ms PCM audio frame with the 4-byte protocol header.
    func sendAudioFrame(_ pcmData: Data) {
        guard state == .active else { return }

        lock.lock()
        let seq = sendSequence
        sendSequence &+= 1
        lock.unlock()

        var frame = Data(capacity: 4 + pcmData.count)
        frame.append(0x01) // direction: client-to-server
        frame.append(0x00) // reserved
        frame.append(UInt8(seq >> 8)) // sequence high byte
        frame.append(UInt8(seq & 0xFF)) // sequence low byte
        frame.append(pcmData)

        webSocket?.send(.data(frame)) { [weak self] error in
            if let error {
                self?.logger.error("Failed to send audio frame: \(error.localizedDescription)")
            }
        }
    }

    // MARK: - Disconnect

    func disconnect() {
        heartbeatTask?.cancel()
        heartbeatTask = nil
        receiveTask?.cancel()
        receiveTask = nil

        webSocket?.cancel(with: .normalClosure, reason: nil)
        webSocket = nil
        urlSession?.invalidateAndCancel()
        urlSession = nil

        lock.lock()
        sendSequence = 0
        lock.unlock()

        setState(.disconnected)
    }

    // MARK: - Receive loop

    private func startReceiving() {
        receiveTask = Task { [weak self] in
            guard let self else { return }
            while !Task.isCancelled {
                guard let ws = self.webSocket else { break }
                do {
                    let message = try await ws.receive()
                    self.handleMessage(message)
                } catch {
                    if !Task.isCancelled {
                        self.logger.error("WebSocket receive error: \(error.localizedDescription)")
                        self.onError?("Connection lost: \(error.localizedDescription)", true)
                        self.setState(.disconnected)
                    }
                    break
                }
            }
        }
    }

    private func handleMessage(_ message: URLSessionWebSocketTask.Message) {
        switch message {
        case let .string(text):
            handleTextMessage(text)
        case let .data(data):
            handleBinaryMessage(data)
        @unknown default:
            logger.warning("Unknown WebSocket message type")
        }
    }

    // MARK: - Text message handling

    private func handleTextMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? String
        else {
            logger.warning("Malformed text message")
            return
        }

        switch type {
        case "auth_result": handleAuthResult(json)
        case "session_started": handleSessionStarted(json)
        case "session_ended": handleSessionEnded(json)
        case "status": handleStatusUpdate(json)
        case "pong": break
        case "error": handleError(json)
        case "fatal_error": handleFatalError(json)
        default: logger.info("Unhandled message type: \(type)")
        }
    }

    private func handleAuthResult(_ json: [String: Any]) {
        if (json["status"] as? String) == "ok" {
            logger.info("Authenticated for practice session")
        } else {
            onError?("Authentication failed", true)
            disconnect()
        }
    }

    private func handleSessionStarted(_ json: [String: Any]) {
        setState(.active)
        sendSequence = 0
        startHeartbeat()
        if let sessionId = json["session_id"] as? String {
            onSessionStarted?(sessionId)
        }
        logger.info("Practice session started")
    }

    private func handleSessionEnded(_ json: [String: Any]) {
        let duration = json["duration_seconds"] as? Int ?? 0
        onSessionEnded?(duration)
        setState(.disconnected)
        logger.info("Practice session ended (\(duration)s)")
    }

    private func handleStatusUpdate(_ json: [String: Any]) {
        if let stateStr = json["state"] as? String,
           let pabloState = PabloState(rawValue: stateStr)
        {
            onPabloStateChanged?(pabloState)
        }
    }

    private func handleError(_ json: [String: Any]) {
        let msg = json["message"] as? String ?? "Unknown error"
        let recoverable = json["recoverable"] as? Bool ?? true
        logger.warning("Practice error: \(msg) (recoverable: \(recoverable))")
        onError?(msg, !recoverable)
    }

    private func handleFatalError(_ json: [String: Any]) {
        let msg = json["message"] as? String ?? "Fatal error"
        logger.error("Fatal practice error: \(msg)")
        onError?(msg, true)
        setState(.disconnected)
    }

    // MARK: - Binary message handling

    private func handleBinaryMessage(_ data: Data) {
        guard data.count >= 4 else {
            logger.warning("Binary frame too short (\(data.count) bytes)")
            return
        }

        // Parse 4-byte header
        let direction = data[0]
        guard direction == 0x02 else {
            logger.warning("Unexpected binary direction byte: \(direction)")
            return
        }

        let flags = data[1]
        let isFinal = (flags & 0x01) != 0
        let pcmData = data.subdata(in: 4 ..< data.count)

        onAudioReceived?(pcmData, isFinal)
    }

    // MARK: - Heartbeat

    private func startHeartbeat() {
        heartbeatTask?.cancel()
        heartbeatTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(15))
                guard !Task.isCancelled else { break }
                let ts = Int(Date().timeIntervalSince1970 * 1000)
                self?.sendJSON(["type": "ping", "ts": ts])
            }
        }
    }

    // MARK: - JSON send helper

    private func sendJSON(_ dict: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: dict),
              let text = String(data: data, encoding: .utf8)
        else { return }

        webSocket?.send(.string(text)) { [weak self] error in
            if let error {
                self?.logger.error("Failed to send JSON: \(error.localizedDescription)")
            }
        }
    }
}

File: windows/PabloCompanion/Services/PracticeWebSocketClient.cs
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;

namespace PabloCompanion.Services;

/// <summary>
/// WebSocket client for practice mode audio streaming.
/// Sends therapist mic audio (PCM 16kHz mono 16-bit) and receives
/// Pablo Bear's response audio (PCM 24kHz mono 16-bit).
/// Uses the hybrid text/binary protocol defined in practice-mode-api.md.
/// </summary>
public sealed class PracticeWebSocketClient : IDisposable
{
    // ── Public state ────────────────────────────────────────────────────────

    public enum ConnectionState { Disconnected, Connecting, Authenticating, WaitingForSession, Active, Ending }
    public enum PabloState { Listening, Processing, Speaking }

    // ── Events ──────────────────────────────────────────────────────────────

    public event Action<ConnectionState>? ConnectionStateChanged;
    public event Action<PabloState>? PabloStateChanged;
    public event Action<byte[], bool>? AudioReceived;
    public event Action<string>? SessionStarted;
    public event Action<int>? SessionEnded;
    public event Action<string, bool>? ErrorOccurred; // message, isFatal

    // ── Properties ──────────────────────────────────────────────────────────

    public ConnectionState State
    {
        get { lock (_lock) return _state; }
    }

    public ushort LastReceivedSequence
    {
        get { lock (_lock) return _lastReceivedSequence; }
    }

    // ── Private fields ──────────────────────────────────────────────────────

    private readonly object _lock = new();
    private ClientWebSocket? _ws;
    private CancellationTokenSource? _cts;
    private ConnectionState _state = ConnectionState.Disconnected;
    private ushort _sendSequence;
    private ushort _lastReceivedSequence;

    // ── Connect ─────────────────────────────────────────────────────────────

    public async Task ConnectAsync(Uri uri)
    {
        Disconnect();

        SetState(ConnectionState.Connecting);

        _cts = new CancellationTokenSource();
        _ws = new ClientWebSocket();

        try
        {
            PabloCompanion.App.Log($"[WS] connecting to {uri}");
            await _ws.ConnectAsync(uri, _cts.Token);
            PabloCompanion.App.Log("[WS] connected, entering authenticating state");
            SetState(ConnectionState.Authenticating);
            _ = Task.Run(() => ReceiveLoopAsync(_cts.Token));
        }
        catch (Exception ex)
        {
            PabloCompanion.App.LogException("[WS] connect failed", ex);
            ErrorOccurred?.Invoke($"Connection failed: {ex.Message}", true);
            SetState(ConnectionState.Disconnected);
        }
    }

    // ── Session lifecycle ───────────────────────────────────────────────────

    public void StartSession(string sessionId)
    {
        SendJson(new { type = "session_start", session_id = sessionId });
        SetState(ConnectionState.WaitingForSession);
    }

    public void ResumeSession(string sessionId, ushort lastSequence)
    {
        SendJson(new { type = "session_resume", session_id = sessionId, last_sequence = (int)lastSequence });
        SetState(ConnectionState.WaitingForSession);
    }

    public void EndSession()
    {
        SendJson(new { type = "session_end" });
        SetState(ConnectionState.Ending);
    }

    public void PauseAudio()
    {
        SendJson(new { type = "audio_pause" });
    }

    public void ResumeAudio()
    {
        SendJson(new { type = "audio_resume" });
    }

    // ── Send audio ──────────────────────────────────────────────────────────

    /// <summary>
    /// Sends a 20ms PCM audio frame with the 4-byte protocol header.
    /// </summary>
    public void SendAudioFrame(byte[] pcmData)
    {
        if (State != ConnectionState.Active) return;

        ushort seq;
        lock (_lock)
        {
            seq = _sendSequence;
            _sendSequence++;
        }

        var frame = new byte[4 + pcmData.Length];
        frame[0] = 0x01; // direction: client-to-server
        frame[1] = 0x00; // flags: reserved
        frame[2] = (byte)(seq >> 8); // sequence high byte
        frame[3] = (byte)(seq & 0xFF); // sequence low byte
        Buffer.BlockCopy(pcmData, 0, frame, 4, pcmData.Length);

        _ = SendBinaryAsync(frame);
    }

    // ── Disconnect ──────────────────────────────────────────────────────────

    public void Disconnect(bool forReconnect = false)
    {
        _cts?.Cancel();
        _cts?.Dispose();
        _cts = null;

        if (_ws?.State == WebSocketState.Open)
        {
            try { _ws.CloseAsync(WebSocketCloseStatus.NormalClosure, null, CancellationToken.None).Wait(1000); }
            catch { /* best-effort close */ }
        }
        _ws?.Dispose();
        _ws = null;

        if (!forReconnect)
        {
            lock (_lock)
            {
                _sendSequence = 0;
                _lastReceivedSequence = 0;
            }
        }

        SetState(ConnectionState.Disconnected);
    }

    public void Dispose() => Disconnect();

    // ── Receive loop ────────────────────────────────────────────────────────

    private async Task ReceiveLoopAsync(CancellationToken ct)
    {
        var buffer = new byte[8192];

        try
        {
            while (!ct.IsCancellationRequested && _ws?.State == WebSocketState.Open)
            {
                using var ms = new MemoryStream();
                WebSocketReceiveResult result;

                do
                {
                    result = await _ws.ReceiveAsync(buffer, ct);
                    ms.Write(buffer, 0, result.Count);
                } while (!result.EndOfMessage);

                if (result.MessageType == WebSocketMessageType.Text)
                {
                    HandleTextMessage(Encoding.UTF8.GetString(ms.ToArray()));
                }
                else if (result.MessageType == WebSocketMessageType.Binary)
                {
                    HandleBinaryMessage(ms.ToArray());
                }
                else if (result.MessageType == WebSocketMessageType.Close)
                {
                    break;
                }
            }
        }
        catch (OperationCanceledException) { /* expected on disconnect */ }
        catch (Exception ex)
        {
            if (!ct.IsCancellationRequested)
            {
                ErrorOccurred?.Invoke($"Connection lost: {ex.Message}", true);
                SetState(ConnectionState.Disconnected);
            }
        }
    }

    // ── Text message handling ───────────────────────────────────────────────

    private void HandleTextMessage(string text)
    {
        PabloCompanion.App.Log($"[WS] recv text: {(text.Length > 200 ? text[..200] + "..." : text)}");
        using var doc = JsonDocument.Parse(text);
        var root = doc.RootElement;

        if (!root.TryGetProperty("type", out var typeProp)) return;
        var type = typeProp.GetString();

        switch (type)
        {
            case "auth_result":
                if (root.TryGetProperty("status", out var status) && status.GetString() != "ok")
                {
                    ErrorOccurred?.Invoke("Authentication failed", true);
                    Disconnect();
                }
                break;

            case "session_started":
                SetState(ConnectionState.Active);
                lock (_lock) _sendSequence = 0;
                StartHeartbeat();
                if (root.TryGetProperty("session_id", out var sid))
                    SessionStarted?.Invoke(sid.GetString()!);
                break;

            case "session_ended":
                var duration = root.TryGetProperty("duration_seconds", out var dur) ? dur.GetInt32() : 0;
                SessionEnded?.Invoke(duration);
                SetState(ConnectionState.Disconnected);
                break;

            case "status":
                if (root.TryGetProperty("state", out var stateProp))
                {
                    var pabloState = stateProp.GetString() switch
                    {
                        "listening" => PabloState.Listening,
                        "processing" => PabloState.Processing,
                        "speaking" => PabloState.Speaking,
                        _ => (PabloState?)null,
                    };
                    if (pabloState.HasValue) PabloStateChanged?.Invoke(pabloState.Value);
                }
                break;

            case "pong":
                break;

            case "error":
                var msg = root.TryGetProperty("message", out var m) ? m.GetString() ?? "Unknown error" : "Unknown error";
                var recoverable = root.TryGetProperty("recoverable", out var r) && r.GetBoolean();
                ErrorOccurred?.Invoke(msg, !recoverable);
                break;

            case "fatal_error":
                var fatalMsg = root.TryGetProperty("message", out var fm) ? fm.GetString() ?? "Fatal error" : "Fatal error";
                ErrorOccurred?.Invoke(fatalMsg, true);
                SetState(ConnectionState.Disconnected);
                break;
        }
    }

    // ── Binary message handling ─────────────────────────────────────────────

    private void HandleBinaryMessage(byte[] data)
    {
        if (data.Length < 4) return;

        var direction = data[0];
        if (direction != 0x02) return; // expect server-to-client

        var flags = data[1];
        var isFinal = (flags & 0x01) != 0;
        var seq = (ushort)((data[2] << 8) | data[3]);

        lock (_lock) _lastReceivedSequence = seq;

        var pcm = new byte[data.Length - 4];
        Buffer.BlockCopy(data, 4, pcm, 0, pcm.Length);

        AudioReceived?.Invoke(pcm, isFinal);
    }

    // ── Heartbeat ───────────────────────────────────────────────────────────

    private void StartHeartbeat()
    {
        _ = Task.Run(async () =>
        {
            while (State == ConnectionState.Active && _cts is { IsCancellationRequested: false })
            {
                await Task.Delay(15_000);
                if (State != ConnectionState.Active) break;
                var ts = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
                SendJson(new { type = "ping", ts });
            }
        });
    }

    // ── Helpers ─────────────────────────────────────────────────────────────

    private void SetState(ConnectionState newState)
    {
        lock (_lock) _state = newState;
        PabloCompanion.App.Log($"[WS] state → {newState}");
        ConnectionStateChanged?.Invoke(newState);
    }

    private void SendJson(object payload)
    {
        if (_ws?.State != WebSocketState.Open) return;
        var json = JsonSerializer.Serialize(payload);
        var bytes = Encoding.UTF8.GetBytes(json);
        _ = SendBinaryAsync(bytes, WebSocketMessageType.Text);
    }

    private async Task SendBinaryAsync(byte[] data, WebSocketMessageType type = WebSocketMessageType.Binary)
    {
        if (_ws?.State != WebSocketState.Open) return;
        try
        {
            await _ws.SendAsync(data, type, true, _cts?.Token ?? CancellationToken.None);
        }
        catch { /* fire-and-forget for audio frames */ }
    }
}
